File: client/lib/api/backtest.ts
/**
 * Backtest API client — POST /api/backtest / POST /api/backtest/reproduce / GET /api/data-freshness.
 *
 * ADR-0027 D1/D2/D3/D4/D5 — PIT rebalance 백테스트 실행.
 * ADR-0033 D2~D6 — 재현 검증 + 데이터 신선도 신뢰성 표시 (M5 #6).
 *
 * 계약 (wire, snake_case):
 *   POST /api/backtest
 *   요청: { pack_slug, pack_version, conditions, start, end, rebalance, cost_commission_bps?, cost_tax_bps? }
 *   응답: { freeze, stats, equity_curve, survivorship_complete, missing_price_ratio, disclaimer_required }
 *
 *   POST /api/backtest/reproduce
 *   요청: { freeze: BacktestFreezeWire, equity_curve: [{date, value}], custom_pack_body?: object|null }
 *   응답: { matches, reproduce_note, pack_tampered, equity_curve }
 *
 *   GET /api/data-freshness
 *   응답: { krx, dart, as_of } — 각 { source, latest_batch_at, is_stale, elapsed_days }
 *
 * No Advice / ADR-0027 D4/D6:
 *   - 결과 타입에 평가/등급/순위 필드 없음 — 사실 통계만.
 *   - 이 모듈은 transport 전용 — 시각 정책 강제는 UI 계층(BacktestPanel).
 *
 * snake↔camel 변환:
 *   - 요청: camelCase 입력 → snake_case wire body.
 *   - 응답: snake_case wire → camelCase 반환.
 *
 * 관련:
 *   - lib/api/client.ts — fetchJson (Authorization Bearer 자동 주입).
 *   - components/Backtest/BacktestPanel.tsx — UI + 시각 게이트.
 */

import { fetchJson } from "./client";
import type { ScreenCondition } from "./screen";

// ── Wire 타입 (snake_case, backend 계약) ─────────────────────────────────────

/** POST /api/backtest 요청 body (snake_case wire). */
interface BacktestBodyWire {
  readonly pack_slug: string;
  readonly pack_version: string;
  readonly conditions: ReadonlyArray<{
    readonly factor: string;
    readonly op: string;
    readonly value: string;
  }>;
  readonly start: string;
  readonly end: string;
  readonly rebalance: "quarterly" | "monthly";
  readonly cost_commission_bps?: number;
  readonly cost_tax_bps?: number;
}

/** freeze artifact wire 구조 (ADR-0027 D5 / ADR-0033 D2 — self-contained 재현 입력). */
interface BacktestFreezeWire {
  readonly result_hash: string;
  readonly pack_content_hash: string;
  // ADR-0033 D2 — pack 정본 재로드 식별자(reproduce 입력).
  readonly pack_slug: string;
  readonly pack_version: string;
  readonly krx_batch_id: string;
  readonly dart_batch_id: string;
  readonly rebalance: "quarterly" | "monthly";
  readonly start: string;
  readonly end: string;
  readonly cost_assumptions: {
    readonly tax_bps: string;
    readonly commission_bps: string;
  };
  // ADR-0033 D2 — self-contained 재현 입력(universe conditions + 전체 data_versions).
  readonly conditions: ReadonlyArray<{
    readonly factor: string;
    readonly op: string;
    readonly value: string;
  }>;
  readonly data_versions: Readonly<Record<string, string>>;
}

/** 통계 wire 구조. Decimal 통계는 string, nullable 통계는 null (외삽 없음). */
interface BacktestStatsWire {
  readonly cagr: string | null;
  readonly cumulative_return: string;
  readonly mdd: string;
  readonly volatility: string | null;
  readonly turnover: string;
}

/** equity curve 포인트 wire. */
interface EquityCurvePointWire {
  readonly date: string;
  readonly value: string;
}

/** POST /api/backtest 응답 wire. */
interface BacktestResultWire {
  readonly freeze: BacktestFreezeWire;
  readonly stats: BacktestStatsWire;
  readonly equity_curve: ReadonlyArray<EquityCurvePointWire>;
  readonly survivorship_complete: boolean;
  readonly missing_price_ratio: string;
  readonly disclaimer_required: boolean;
}

// ── 공개 타입 (camelCase) ────────────────────────────────────────────────────

/** freeze artifact — ADR-0027 D5 (재현 불변식) / ADR-0033 D2 (self-contained 재현 입력). */
export interface BacktestFreeze {
  readonly resultHash: string;
  readonly packContentHash: string;
  /** ADR-0033 D2 — pack 정본 재로드 식별자(reproduce 입력). */
  readonly packSlug: string;
  readonly packVersion: string;
  readonly krxBatchId: string;
  readonly dartBatchId: string;
  readonly rebalance: "quarterly" | "monthly";
  readonly start: string;
  readonly end: string;
  /** 거래비용 가정 — D2 강제 노출. 숨기지 않음. */
  readonly costAssumptions: {
    readonly taxBps: string;
    readonly commissionBps: string;
  };
  /** ADR-0033 D2 — self-contained 재현 입력(universe conditions, canonical 정렬). */
  readonly conditions: ReadonlyArray<{
    readonly factor: string;
    readonly op: string;
    readonly value: string;
  }>;
  /** ADR-0033 D2 — 정책+batch_id 전체 freeze(재실행 데이터 시점 재현). */
  readonly dataVersions: Readonly<Record<string, string>>;
}

/**
 * 백테스트 통계 — 사실 통계만, 평가/등급/순위 필드 없음 (ADR-0027 D6).
 *
 * - Decimal 통계: string (정밀도 보존).
 * - nullable 통계: null (외삽 없음 — CAGR, volatility 는 기간/종목 부족 시 null).
 */
export interface BacktestStats {
  readonly cagr: string | null;
  readonly cumulativeReturn: string;
  readonly mdd: string;
  readonly volatility: string | null;
  readonly turnover: string;
}

/** equity curve 포인트 (camelCase). */
export interface EquityCurvePoint {
  readonly date: string;
  readonly value: string;
}

/**
 * 백테스트 결과 — ADR-0027 D4/D6 출력 게이트.
 *
 * 포함: freeze, stats, equityCurve, survivorshipComplete, missingPriceRatio, disclaimerRequired.
 * 제외: 평가/등급/순위/우수/별점 필드 일절 없음 (D6 결과 = 사실 통계만).
 */
export interface BacktestResult {
  readonly freeze: BacktestFreeze;
  readonly stats: BacktestStats;
  readonly equityCurve: ReadonlyArray<EquityCurvePoint>;
  /** survivorship bias 가능 여부 (ADR-0027 D3). */
  readonly survivorshipComplete: boolean;
  /** 가격 누락 비율 string (0 이상 1 미만). */
  readonly missingPriceRatio: string;
  /** 과거성과 디스클레이머 게이트 필요 여부 (ADR-0027 D4). */
  readonly disclaimerRequired: boolean;
}

/** runBacktest 입력 파라미터 (camelCase). */
export interface RunBacktestParams {
  readonly packSlug: string;
  readonly packVersion: string;
  readonly conditions: ReadonlyArray<ScreenCondition>;
  readonly start: string;
  readonly end: string;
  readonly rebalance: "quarterly" | "monthly";
  readonly costCommissionBps?: number;
  readonly costTaxBps?: number;
}

// ── snake↔camel 변환 ─────────────────────────────────────────────────────────

/** wire freeze → camelCase BacktestFreeze. */
function mapFreeze(w: BacktestFreezeWire): BacktestFreeze {
  return {
    resultHash: w.result_hash,
    packContentHash: w.pack_content_hash,
    packSlug: w.pack_slug,
    packVersion: w.pack_version,
    krxBatchId: w.krx_batch_id,
    dartBatchId: w.dart_batch_id,
    rebalance: w.rebalance,
    start: w.start,
    end: w.end,
    costAssumptions: {
      taxBps: w.cost_assumptions.tax_bps,
      commissionBps: w.cost_assumptions.commission_bps,
    },
    // ADR-0033 D2 — self-contained 재현 입력 운반(읽기 전용 그대로).
    conditions: w.conditions,
    dataVersions: w.data_versions,
  };
}

/** wire stats → camelCase BacktestStats. */
function mapStats(w: BacktestStatsWire): BacktestStats {
  return {
    cagr: w.cagr,
    cumulativeReturn: w.cumulative_return,
    mdd: w.mdd,
    volatility: w.volatility,
    turnover: w.turnover,
  };
}

/** wire equity_curve → camelCase EquityCurvePoint[]. */
function mapEquityCurve(
  w: ReadonlyArray<EquityCurvePointWire>,
): ReadonlyArray<EquityCurvePoint> {
  return w.map((pt) => ({ date: pt.date, value: pt.value }));
}

/** wire 응답 → BacktestResult (camelCase). */
function mapResult(w: BacktestResultWire): BacktestResult {
  return {
    freeze: mapFreeze(w.freeze),
    stats: mapStats(w.stats),
    equityCurve: mapEquityCurve(w.equity_curve),
    survivorshipComplete: w.survivorship_complete,
    missingPriceRatio: w.missing_price_ratio,
    disclaimerRequired: w.disclaimer_required,
  };
}

// ── Reproduce wire 타입 (ADR-0033 D2~D5) ─────────────────────────────────────

/** POST /api/backtest/reproduce 요청 body (snake_case wire). */
interface BacktestReproduceBodyWire {
  readonly freeze: BacktestFreezeWire;
  readonly equity_curve: ReadonlyArray<EquityCurvePointWire>;
  readonly custom_pack_body?: object | null;
}

/** POST /api/backtest/reproduce 응답 wire. */
interface BacktestReproduceResultWire {
  readonly matches: boolean;
  readonly reproduce_note: string | null;
  readonly pack_tampered: boolean;
  readonly equity_curve: ReadonlyArray<EquityCurvePointWire>;
}

// ── DataFreshness wire 타입 (ADR-0033 D6) ─────────────────────────────────────

/** GET /api/data-freshness 소스별 신선도 wire. */
interface SourceFreshnessWire {
  readonly source: string;
  readonly latest_batch_at: string | null;
  readonly is_stale: boolean;
  readonly elapsed_days: number | null;
}

/** GET /api/data-freshness 응답 wire. */
interface DataFreshnessWire {
  readonly krx: SourceFreshnessWire;
  readonly dart: SourceFreshnessWire;
  readonly kosis: SourceFreshnessWire;
  readonly as_of: string;
}

// ── 공개 타입 (ADR-0033) ─────────────────────────────────────────────────────

/**
 * 백테스트 재현 결과 — ADR-0033 D3/D4/D5 (사실만, 판단 없음).
 *
 * - matches: 재현 equity curve 가 원본과 일치하는지 여부(사실).
 * - reproduceNote: 불일치 원인 메모 (있으면).
 * - packTampered: pack content_hash 불일치 — 변조 또는 pack 교체 가능성(사실).
 * - equityCurve: 재현 실행의 equity curve.
 */
export interface BacktestReproduceResult {
  readonly matches: boolean;
  readonly reproduceNote: string | null;
  readonly packTampered: boolean;
  readonly equityCurve: ReadonlyArray<EquityCurvePoint>;
}

/**
 * 소스별 데이터 신선도 — ADR-0033 D6 (사실만).
 *
 * - source: 소스 식별자 (예: "krx", "dart").
 * - latestBatchAt: 최신 배치 기준일 (ISO 8601, 없으면 null).
 * - isStale: 갱신 지연 여부 (사실).
 * - elapsedDays: 마지막 배치로부터 경과 일수 (없으면 null).
 */
export interface SourceFreshness {
  readonly source: string;
  readonly latestBatchAt: string | null;
  readonly isStale: boolean;
  readonly elapsedDays: number | null;
}

/**
 * 데이터 신선도 — ADR-0033 D6 / ADR-0036 D7 (KRX/DART/KOSIS 각 소스 신선도 + 기준 시점).
 */
export interface DataFreshness {
  readonly krx: SourceFreshness;
  readonly dart: SourceFreshness;
  readonly kosis: SourceFreshness;
  readonly asOf: string;
}

// ── snake↔camel: reproduce / freshness ───────────────────────────────────────

/**
 * BacktestFreeze(camel) → BacktestFreezeWire(snake) 역변환.
 * POST /api/backtest/reproduce 요청 body 조립용.
 */
export function freezeToWire(f: BacktestFreeze): BacktestFreezeWire {
  return {
    result_hash: f.resultHash,
    pack_content_hash: f.packContentHash,
    pack_slug: f.packSlug,
    pack_version: f.packVersion,
    krx_batch_id: f.krxBatchId,
    dart_batch_id: f.dartBatchId,
    rebalance: f.rebalance,
    start: f.start,
    end: f.end,
    cost_assumptions: {
      tax_bps: f.costAssumptions.taxBps,
      commission_bps: f.costAssumptions.commissionBps,
    },
    conditions: f.conditions,
    data_versions: f.dataVersions,
  };
}

/** wire reproduce 응답 → camelCase BacktestReproduceResult. */
function mapReproduceResult(w: BacktestReproduceResultWire): BacktestReproduceResult {
  return {
    matches: w.matches,
    reproduceNote: w.reproduce_note,
    packTampered: w.pack_tampered,
    equityCurve: mapEquityCurve(w.equity_curve),
  };
}

/** wire SourceFreshness → camelCase SourceFreshness. */
function mapSourceFreshness(w: SourceFreshnessWire): SourceFreshness {
  return {
    source: w.source,
    latestBatchAt: w.latest_batch_at,
    isStale: w.is_stale,
    elapsedDays: w.elapsed_days,
  };
}

// ── 공개 함수 ─────────────────────────────────────────────────────────────────

/**
 * POST /api/backtest — PIT rebalance 백테스트 실행.
 *
 * @param params  RunBacktestParams (camelCase)
 * @param signal  AbortController signal (선택)
 * @returns BacktestResult (camelCase)
 * @throws ApiError — 401 미인증, 404 팩 없음, 422 검증 실패.
 */
export async function runBacktest(
  params: RunBacktestParams,
  signal?: AbortSignal,
): Promise<BacktestResult> {
  const body: BacktestBodyWire = {
    pack_slug: params.packSlug,
    pack_version: params.packVersion,
    conditions: params.conditions,
    start: params.start,
    end: params.end,
    rebalance: params.rebalance,
    ...(params.costCommissionBps !== undefined
      ? { cost_commission_bps: params.costCommissionBps }
      : {}),
    ...(params.costTaxBps !== undefined
      ? { cost_tax_bps: params.costTaxBps }
      : {}),
  };

  const wire = await fetchJson<BacktestResultWire>("/api/backtest", {
    method: "POST",
    body,
    signal,
  });

  return mapResult(wire);
}

/**
 * POST /api/backtest/reproduce — 백테스트 재현 검증 (ADR-0033 D2~D5).
 *
 * freeze + equity_curve 를 서버에 전달해 동일 결과를 재현 가능한지 검증.
 * matches=true/false, packTampered=true/false 는 사실 값 — 판단 아님.
 *
 * @param freeze BacktestFreeze (BacktestResult.freeze 그대로)
 * @param equityCurve BacktestResult.equityCurve 그대로
 * @param customPackBody custom pack body (선택, null 이면 미전달)
 * @param signal AbortController signal
 * @returns BacktestReproduceResult (camelCase)
 * @throws ApiError — 422 검증 실패 등.
 */
export async function reproduceBacktest(
  freeze: BacktestFreeze,
  equityCurve: ReadonlyArray<EquityCurvePoint>,
  customPackBody?: object | null,
  signal?: AbortSignal,
): Promise<BacktestReproduceResult> {
  const body: BacktestReproduceBodyWire = {
    freeze: freezeToWire(freeze),
    equity_curve: equityCurve.map((pt) => ({ date: pt.date, value: pt.value })),
    ...(customPackBody !== undefined ? { custom_pack_body: customPackBody } : {}),
  };

  const wire = await fetchJson<BacktestReproduceResultWire>("/api/backtest/reproduce", {
    method: "POST",
    body,
    signal,
  });

  return mapReproduceResult(wire);
}

/**
 * GET /api/data-freshness — KRX/DART 데이터 신선도 조회 (ADR-0033 D6).
 *
 * 응답은 사실 값 — latestBatchAt(날짜), isStale(경과 여부), elapsedDays(경과 일수).
 * 판단·등락 어휘 없음.
 *
 * @param signal AbortController signal
 * @returns DataFreshness (camelCase)
 */
export async function getDataFreshness(
  signal?: AbortSignal,
): Promise<DataFreshness> {
  const wire = await fetchJson<DataFreshnessWire>("/api/data-freshness", {
    method: "GET",
    signal,
  });

  return {
    krx: mapSourceFreshness(wire.krx),
    dart: mapSourceFreshness(wire.dart),
    kosis: mapSourceFreshness(wire.kosis),
    asOf: wire.as_of,
  };
}

File: server/app/schemas/backtest.py
"""Backtest endpoint wire schemas — POST /api/backtest (ADR-0027).

설계 (screen.py 의 ConditionIn / strict 패턴 미러):
- BacktestConditionIn strict — factor canonical_id regex (text injection 차단).
- 응답은 wire snake_case — frontend(Backtest/) 가 그대로 의존.
- disclaimer_required: true 고정 — 과거성과 디스클레이머 게이트 (ADR-0027 D4).
- survivorship_complete / missing_price_ratio — survivorship 디스클로저 (D3).
- 평가 라벨/등급 0 — 사실 통계만 (D6).
"""

from __future__ import annotations

from datetime import date
from enum import Enum
from typing import Final, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.screen import OpEnum

__all__ = [
    "BacktestConditionIn",
    "BacktestReproduceIn",
    "BacktestReproduceOut",
    "BacktestRequestIn",
    "BacktestResultOut",
    "CostAssumptionsOut",
    "EquityPointOut",
    "FreezeOut",
    "RebalanceEnum",
    "StatsOut",
]

_STRICT_MODEL_CONFIG = ConfigDict(strict=True, extra="forbid", frozen=True)

# screen.py 와 동일 factor canonical_id 형식 — text injection 차단.
_FACTOR_ID_PATTERN: Final[str] = r"^[a-z][a-z0-9-]*(:[a-z][a-z0-9-]*)+$"
_VALUE_MAX_LENGTH: Final[int] = 200

# pack 식별자 형식 — screen.py CustomScreenRunQueryIn 와 일관.
_PACK_SLUG_PATTERN: Final[str] = r"^[a-z][a-z0-9-]*(/[a-z0-9][a-z0-9-]*)?$"
_PACK_VERSION_PATTERN: Final[str] = r"^[0-9]+\.[0-9]+\.[0-9]+$"

# 거래비용 bps 상한 — 비현실적 입력 차단 (1000 bps = 10%). ADR-0027 D2 는 0 으로
# 숨길 수 없으나 0 입력 자체는 허용(명시 노출되므로). 음수 거부.
_MAX_COST_BPS: Final[float] = 1000.0


class RebalanceEnum(str, Enum):
    """rebalance 주기 — 분기말 / 월말 (ADR-0027 D1)."""

    QUARTERLY = "quarterly"
    MONTHLY = "monthly"


class BacktestConditionIn(BaseModel):
    """백테스트 universe 필터 조건 — screen.py ConditionIn 동형.

    Attributes:
        factor: factor canonical_id (regex 강제).
        op: OpEnum 연산자.
        value: 비교값 (str — 호출자 Decimal 변환).
    """

    model_config = ConfigDict(strict=False, extra="forbid", frozen=True)

    factor: str = Field(
        pattern=_FACTOR_ID_PATTERN, min_length=3, max_length=200,
        description="Factor canonical_id (예: 'per:ttm-consolidated-ifrs').",
    )
    op: OpEnum
    value: str = Field(min_length=1, max_length=_VALUE_MAX_LENGTH)


class BacktestRequestIn(BaseModel):
    """POST /api/backtest body — pack 식별자 + 기간 + 주기 + (옵션) 거래비용.

    pack 은 registry.resolve(slug, version, user_id=current_user, ...) 로 해소 —
    빌트인/reference/custom 모두. start <= end 강제.

    Attributes:
        pack_slug / pack_version: pack 식별자.
        conditions: universe 필터 조건 (1~32 개).
        start / end: 백테스트 기간.
        rebalance: "quarterly" | "monthly".
        cost_commission_bps / cost_tax_bps: 거래비용 override (옵션). None 이면
            엔진 default (DEFAULT_COMMISSION_BPS / DEFAULT_TAX_BPS). ADR-0027 D2 —
            0 으로 숨길 수 없으나 결과에 명시 노출되므로 0 입력 허용.

    Note (strict=False):
        date / RebalanceEnum 은 JSON wire 문자열("2022-01-01" / "quarterly") 로
        들어오므로 coercion 을 위해 non-strict. `extra="forbid"` + `frozen=True`
        는 유지 — 미지 키 차단 + immutability (ConditionIn 패턴 동형).
    """

    model_config = ConfigDict(strict=False, extra="forbid", frozen=True)

    pack_slug: str = Field(
        pattern=_PACK_SLUG_PATTERN, min_length=1, max_length=128,
    )
    pack_version: str = Field(
        pattern=_PACK_VERSION_PATTERN, min_length=5, max_length=32,
    )
    conditions: list[BacktestConditionIn] = Field(min_length=1, max_length=32)
    start: date
    end: date
    rebalance: RebalanceEnum
    cost_commission_bps: float | None = Field(
        default=None, ge=0.0, le=_MAX_COST_BPS,
    )
    cost_tax_bps: float | None = Field(default=None, ge=0.0, le=_MAX_COST_BPS)

    @field_validator("end")
    @classmethod
    def _validate_period(cls, v: date, info) -> date:  # type: ignore[no-untyped-def]
        start = info.data.get("start")
        if start is not None and v < start:
            raise ValueError("end 가 start 보다 빠릅니다 (start <= end 강제).")
        return v


# =============================================================================
# 응답 — wire snake_case (frontend 의존)
# =============================================================================

class CostAssumptionsOut(BaseModel):
    """적용된 거래비용 가정 — ADR-0027 D2 강제 노출."""

    model_config = _STRICT_MODEL_CONFIG

    tax_bps: str
    commission_bps: str


class FreezeOut(BaseModel):
    """freeze artifact wire — self-contained 재현 입력 (ADR-0027 D5 / ADR-0033 D2).

    M5 #3a-0: backtest reproduce(예정)를 위한 self-contained export. result_hash
    식별자뿐 아니라 **재실행에 필요한 전 입력**(pack 식별·universe conditions·
    data_versions)을 운반한다(screen `ScreenRunSnapshotOut` 패턴). client 가 이
    freeze + equity_curve(baseline)를 보관해 reproduce 요청에 그대로 사용한다.

    **불변식(ADR-0033 D3)**: 본 wire 필드 추가는 `result_hash` 입력
    (`BacktestSnapshotBuilder.build` 의 hash_input)과 **무관** — 기존 frozen
    backtest 의 result_hash 가 byte-동일하게 유지된다(frozen 재현 호환).

    Note (strict=False):
        응답 직렬화(BacktestResultOut.freeze)뿐 아니라 **reproduce 요청 입력**
        (BacktestReproduceIn.freeze)으로도 재사용된다. 입력 시 date/Literal/tuple
        이 JSON wire 문자열로 들어오므로 coercion 을 위해 non-strict. `extra=forbid`
        + `frozen=True` 는 유지 — 미지 키 차단 + immutability(self-contained 무결성).
    """

    model_config = ConfigDict(strict=False, extra="forbid", frozen=True)

    result_hash: str
    pack_content_hash: str
    # ADR-0033 D2 — pack 정본 재로드 식별자(reproduce 시 PackRegistry.resolve 입력).
    pack_slug: str
    pack_version: str
    krx_batch_id: str
    dart_batch_id: str
    rebalance: Literal["quarterly", "monthly"]
    start: date
    end: date
    cost_assumptions: CostAssumptionsOut
    # ADR-0033 D2 — self-contained 재현 입력. universe 필터 conditions(canonical
    # 정렬 — snapshot 과 동일 순서) + 전체 data_versions(정책+batch_id). 재실행이
    # 같은 모집단·데이터 시점을 재현하도록 운반(snapshot.conditions/data_versions).
    conditions: tuple[dict[str, str], ...]
    data_versions: dict[str, str]
    # ADR-0033 D3 / M1 — 산식 세대 식별자(BACKTEST_ENGINE_VERSION). reproduce 시
    # 현재 엔진 버전과 비교해 "산식 세대 bump(예: 버그 수정)로 인한 benign 불일치"를
    # 데이터 드리프트/변조와 구분(engine_version_superseded 진단). result_hash 입력엔
    # 이미 포함(snapshot hash_input) — 본 wire 노출은 hash 무영향(D3 불변식).
    # Optional(default None): 본 필드 도입 **이전**에 client 가 보관한 구 freeze 에는
    # 부재하므로 None 으로 파싱(하위호환). None 이면 세대 비교 불가 → 기존 동작.
    engine_version: str | None = None


class StatsOut(BaseModel):
    """사실 통계 — 평가 라벨/등급 0 (ADR-0027 D6).

    Decimal 은 str 로 직렬화 (정밀도 보존 — float drift 회피). None 가능 통계는
    null (외삽 금지 — 계산 불가 시 사실대로 null).
    """

    model_config = _STRICT_MODEL_CONFIG

    cagr: str | None
    cumulative_return: str
    mdd: str
    volatility: str | None
    turnover: str


class EquityPointOut(BaseModel):
    """equity curve 단일 점 — 날짜별 누적가치 (초기 1.0).

    Note (strict=False): 응답 직렬화뿐 아니라 reproduce 요청의 baseline
    equity_curve(BacktestReproduceIn.equity_curve) 입력으로도 재사용된다. date 가
    JSON wire 문자열로 들어오므로 coercion 을 위해 non-strict. extra=forbid +
    frozen 은 유지 (FreezeOut 와 동일 dual-purpose 사유).
    """

    model_config = ConfigDict(strict=False, extra="forbid", frozen=True)

    date: date
    value: str


class BacktestReproduceIn(BaseModel):
    """POST /api/backtest/reproduce body — frozen freeze + baseline equity_curve.

    screen `ReproduceIn` 의 backtest 등가물 (ADR-0033 D2~D4). self-contained freeze
    (`FreezeOut` — pack 식별·conditions·data_versions·기간·주기·거래비용) + 그와
    함께 보관된 baseline equity_curve 를 받아 frozen batch 기준 재실행 + 동일 검증.

    custom run 은 export 동봉 봉인 pack body(`custom_pack_body`)로 자기완결 재현
    (user/DB 무관 — ADR-0021 D5). 빌트인/reference run 은 None (registry 가 frozen
    slug/version 으로 정본 재로드).

    Note (strict=False):
        freeze 내부의 date / Literal 등이 JSON wire 문자열로 들어오므로 coercion 을
        위해 non-strict. `extra="forbid"` + `frozen=True` 유지 (BacktestRequestIn 동형).
    """

    model_config = ConfigDict(strict=False, extra="forbid", frozen=True)

    freeze: FreezeOut
    equity_curve: tuple[EquityPointOut, ...]
    custom_pack_body: dict | None = None


class BacktestReproduceOut(BaseModel):
    """POST /api/backtest/reproduce 응답 — 재현 동일 여부 + 재실행 equity_curve.

    Attributes:
        matches: 재실행 equity_curve == baseline (byte-동일 재현 성공). False 면
            frozen batch 이후 데이터/정책 변화 또는 재현 불가 (reproduce_note 참조).
        reproduce_note: 재현 결과 안내 — 성공/불일치/재로드 실패 사유.
        equity_curve: frozen batch cutoff 로 재실행한 equity curve (날짜별 누적가치).
    """

    model_config = _STRICT_MODEL_CONFIG

    matches: bool
    reproduce_note: str | None
    equity_curve: tuple[EquityPointOut, ...]
    # pack_tampered: pack 변조(content_hash 불일치)로 재현 불가인 경우만 True.
    # BacktestReproduceResult 와 동형(ADR-0033/M5 #3b). default False(하위호환).
    pack_tampered: bool = False
    # engine_version_superseded: freeze 의 산식 세대(engine_version)가 현재
    # BACKTEST_ENGINE_VERSION 과 다를 때 True(M1). matches 와 **독립** 신호 —
    # 분할 없는 backtest 는 세대가 달라도 equity_curve 가 byte-동일(matches=True)할
    # 수 있다. matches=False 와 함께면 "데이터 변조/드리프트가 아닌 산식 수정으로
    # 인한 benign 불일치"임을 client/운영이 구분(M5 신선도 진단 false-alarm 방지).
    # 구 freeze(engine_version 부재)는 비교 불가 → False(하위호환).
    engine_version_superseded: bool = False


class BacktestResultOut(BaseModel):
    """POST /api/backtest 응답 — 사실 통계 + survivorship 디스클로저 + 디스클레이머 게이트.

    Attributes:
        freeze: 재현 식별자 (result_hash / pack_content_hash / batch_id / 기간 /
            주기 / 거래비용 가정).
        stats: 사실 통계 (cagr / cumulative_return / mdd / volatility / turnover).
        equity_curve: 날짜별 누적가치.
        survivorship_complete: 가격 누락 0 이면 True. False 면 UI 가 survivorship
            bias 경고 강제 (ADR-0027 D3).
        missing_price_ratio: survivorship 측정 — 가격 누락 비율 (0~1).
        disclaimer_required: 항상 true — 과거성과 디스클레이머 게이트 (ADR-0027
            D4). disclaimer 없는 백테스트 렌더 차단 (ADR-0007 D8.1).
    """

    model_config = _STRICT_MODEL_CONFIG

    freeze: FreezeOut
    stats: StatsOut
    equity_curve: tuple[EquityPointOut, ...]
    survivorship_complete: bool
    missing_price_ratio: str
    disclaimer_required: Literal[True] = True
